CMark.hsc

```diff
@@ -45,41 +51,82 @@ import Data.Text.Encoding (encodeUtf8)
-commonmarkToHtml :: [CMarkOption] -> Text -> Text
-commonmarkToHtml opts = commonmarkToX render_html opts Nothing
-  where render_html n o _ = c_cmark_render_html n o
+commonmarkToHtml :: [CMarkOption] -> [CMarkExtension] -> Text -> Text
+commonmarkToHtml opts exts =
+  commonmarkToX render_html opts exts Nothing
+  where exts' = Unsafe.unsafePerformIO $ fmap catMaybes $ mapM resolveExt exts
+        render_html n o _ = Unsafe.unsafePerformIO $ do
+          llist <- extsToLlist exts'
+          let r = c_cmark_render_html n o llist
+          freeLlist llist
+          return r
 
 -- | Convert CommonMark formatted text to CommonMark XML, using cmark's
 -- built-in renderer.
-commonmarkToXml :: [CMarkOption] -> Text -> Text
-commonmarkToXml opts = commonmarkToX render_xml opts Nothing
+commonmarkToXml :: [CMarkOption] -> [CMarkExtension] -> Text -> Text
+commonmarkToXml opts exts = commonmarkToX render_xml opts exts Nothing
   where render_xml n o _ = c_cmark_render_xml n o
 
 -- | Convert CommonMark formatted text to groff man, using cmark's
 -- built-in renderer.
-commonmarkToMan :: [CMarkOption] -> Maybe Int -> Text -> Text
+commonmarkToMan :: [CMarkOption] -> [CMarkExtension] -> Maybe Int -> Text -> Text
 commonmarkToMan = commonmarkToX c_cmark_render_man
 
 -- | Convert CommonMark formatted text to latex, using cmark's
 -- built-in renderer.
-commonmarkToLaTeX :: [CMarkOption] -> Maybe Int -> Text -> Text
+commonmarkToLaTeX :: [CMarkOption] -> [CMarkExtension] -> Maybe Int -> Text -> Text
 commonmarkToLaTeX = commonmarkToX c_cmark_render_latex
 
 -- | Convert CommonMark formatted text to a structured 'Node' tree,
 -- which can be transformed or rendered using Haskell code.
-commonmarkToNode :: [CMarkOption] -> Text -> Node
-commonmarkToNode opts s = Unsafe.unsafePerformIO $ do
-  nptr <- TF.withCStringLen s $! \(ptr, len) ->
-             c_cmark_parse_document ptr len (combineOptions opts)
+commonmarkToNode :: [CMarkOption] -> [CMarkExtension] -> Text -> Node
+commonmarkToNode opts exts s = Unsafe.unsafePerformIO $ do
+  exts' <- fmap catMaybes $ mapM resolveExt exts
+  parser <- c_cmark_parser_new (combineOptions opts)
+  mapM_ (c_cmark_parser_attach_syntax_extension parser) exts'
+  TF.withCStringLen s $! \(ptr, len) ->
+             c_cmark_parser_feed parser ptr len
+  nptr <- c_cmark_parser_finish parser
+  c_cmark_parser_free parser
   fptr <- newForeignPtr c_cmark_node_free nptr
   withForeignPtr fptr toNode
 
-nodeToHtml :: [CMarkOption] -> Node -> Text
-nodeToHtml opts = nodeToX render_html opts Nothing
-  where render_html n o _ = c_cmark_render_html n o
+nodeToHtml :: [CMarkOption] -> [CMarkExtension] -> Node -> Text
+nodeToHtml opts exts =
+  nodeToX render_html opts Nothing
+  where exts' = Unsafe.unsafePerformIO $ fmap catMaybes $ mapM resolveExt exts
+        render_html n o _ = Unsafe.unsafePerformIO $ do
+          llist <- extsToLlist exts'
+          let r = c_cmark_render_html n o llist
+          freeLlist llist
```

Comment:
Would've preferred to use `newForeignPtr` for `llist` here, but it requires a `FunPtr` for the first argument (i.e. something directly `ccall`'able) and we need to provide a partially-applied function, per `freeLlist`'s definition. I don't know if there's anything much neater to be done here.